src/app/login.service.spec.ts

```diff
@@ -0,0 +1,42 @@
+import {HttpClientTestingModule} from '@angular/common/http/testing';
+import {TestBed} from '@angular/core/testing';
+
+import {LoginService} from './login.service';
+import {HttpStalkerService} from './http-stalker.service';
+import {of} from 'rxjs';
+import {HttpResponse, HttpHeaders} from '@angular/common/http';
+import {User} from './user';
+
+describe('LoginService', () => {
+  let service: LoginService;
+  const httpStalker = jasmine.createSpyObj('HttpStalkerService', ['fakepost']);
+  const httpSpy = httpStalker.fakepost.and.returnValue(
+    of(new HttpResponse({body: null, headers: new HttpHeaders(), status: 200})),
+  );
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      providers: [{provide: HttpStalkerService, useValue: httpStalker}],
+    });
+    service = TestBed.inject(LoginService);
+  });
+
+  it('should be created', () => {
+    expect(service).toBeTruthy();
+  });
+  it('should call httpStalker service', () => {
+    service.login(new User());
+    expect(httpSpy.calls.any()).toBe(true, 'fakepost called');
+  });
+  /* it('should handle errors', () => {
+    httpSpy = httpStalker.fakepost.and.returnValue(
+      of(new HttpErrorResponse({error: null})),
+    );
+    const errorspy = jasmine.createSpy('handleError');
+    service.login(new User());
+    expect(errorspy.calls.any()).toBeTrue();
+     expect(service.login(new User()).subscribe()).toEqual(
+      new HttpResponse({body: null, headers: new HttpHeaders(), status: 400}),
+    );
```

Comment:
questi (che sono test di unità) non dovrebbero MAI fare delle post a servizi reali (NEANCHE IL MOCK DI PRISM) perché così hanno una dipendenza "nascosta" da un servizio esterno. ci sono due fasi diverse per fare bene i test:

1. assicurarsi che il codice sia consistente con se stesso (faccia le cose che penso che faccia)
2. assicurarsi che il codice funzioni nel mondo reale. questo è compito o dell'e2e o semplicemente di "vedere se funziona" con il mock di prism.

Reply:
infatti non fanno mai call a servizi/componenti diversi da loro stessi, usiamo sempre un mock del servizio o più specificamente della funzione e tramite le spies di jasmine andiamo a definire noi cosa fanno quelle funzioni invece di essere chiamate veramente, e andiamo anche a verificare se effettivamente sono state chiamate.

Reply:
> infatti non fanno mai call a servizi/componenti diversi da loro stessi, usiamo sempre un mock del servizio o più specificamente della funzione e tramite le spies di jasmine andiamo a definire noi cosa fanno quelle funzioni invece di essere chiamate veramente, e andiamo anche a verificare se effettivamente sono state chiamate.

si le spies sono fighissime; su java tocca importare una libreria apposta solo per avere quella funzionalità. 

> usiamo sempre un mock del servizio

tranne che del servizio che stai attualmente testando, no? cioè, in login.service.spec.ts tutti i service sono dei mock tranne login.service

Reply:
si si a parte quello che sto testando